lib/resty/wasmx/shm.lua

```
@@ -295,78 +320,186 @@ local function set_kv_value(zone, key, value, cas)
-        return { type = "counter", value = tonumber(cmetric.slots[0].counter) }
+        return {
+            type = "counter",
+            value = tonumber(cmetric.slots[0].counter),
+        }
+    end
 
-    elseif cmetric.metric_type == _types.ffi_metric.GAUGE then
-        return { type = "gauge", value = tonumber(cmetric.slots[0].gauge.value) }
+    if cmetric.metric_type == _types.ffi_metric.GAUGE then
+        return {
+            type = "gauge",
+            value = tonumber(cmetric.slots[0].gauge.value),
+        }
+    end
 
-    elseif cmetric.metric_type == _types.ffi_metric.HISTOGRAM then
+    if cmetric.metric_type == _types.ffi_metric.HISTOGRAM then
         local hbuf = cmetric.slots[0].histogram
         local ch = ffi_cast("ngx_wa_metrics_histogram_t *", hbuf)
         local h = { type = "histogram", value = {} }
 
         for i = 0, ch.n_bins do
             local cb = ch.bins[i]
-
             if cb.upper_bound == 0 then
                 break
             end
 
-            h.value[#(h.value) + 1] = { ub = cb.upper_bound, count = cb.count }
+            insert(h.value, {
+                ub = cb.upper_bound,
+                count = cb.count,
+            })
         end
 
         return h
     end
+
+    assert(false, "unreachable")
 end
 
 
-local function get_metric(zone, metric_id)
+local function metrics_get_by_id(zone, metric_id)
     if type(metric_id) ~= "number" then
         error("metric_id must be a number", 2)
     end
 
     ffi_fill(_mbuf, _mbs)
     ffi_fill(_hbuf, _hbs)
 
-    local rc = C.ngx_wa_ffi_shm_get_metric(metric_id, nil,
-                                           _mbuf, _mbs, _hbuf, _hbs)
+    local rc = C.ngx_wa_ffi_shm_metric_get(metric_id, nil,
+                                           _mbuf, _mbs,
+                                           _hbuf, _hbs)
     if rc == FFI_DECLINED then
-        return nil, "metric not found"
+        return nil
+    end
```

We should return an error message `metric not found`.

---

Got it. Thanks for the explanation!